File: src/python/1012/main.py
# Date : 2017.11.16
# Author : seungyeon choi
# 1012 (유기농배추)

import sys
dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]
for _ in range(int(sys.stdin.readline())):
    ans = 0
    M, N, K = map(int, sys.stdin.readline().split());
    graph = [[0]*M for _ in range(N)]
    for _ in range(K):
        X, Y = map(int, sys.stdin.readline().split());
        graph[Y][X] = 1
    for i in range(N):
        for j in range(M):
            q = []
            if graph[i][j] :
                ans += 1
                q.append((i, j))
                while q :
                    x, y = q.pop(0)
                    graph[x][y] = 0
                    for k in range(4):
                        tx, ty = x + dx[k], y + dy[k]
                        if 0<=tx<N and 0<=ty<M and graph[tx][ty]:
                            q.append((tx, ty))
                            graph[tx][ty] = 0
    print (ans)

File: src/c++/1012/main.cc
/* Date : 2019.4.25
 * Author : seungyeon choi
 * 1012 (유기농 배추)
 */

#include <cstdio>

int graph[52][52];
int dx[4] = {1, -1, 0, 0};
int dy[4] = {0, 0, 1, -1};

int t, m, n, k, x, y, ans;

void dfs(int i, int j) {
  graph[i][j] = 0;
  for (int k = 0; k < 4; k++)
    if (graph[i + dx[k]][j + dy[k]])
      dfs(i + dx[k], j + dy[k]);
}

int main() {
  scanf("%d", &t);

  for (int i = 0; i < t; i++) {
    scanf("%d %d %d", &m, &n, &k);

    for (int j = 0; j < k; j++) {
      scanf("%d %d", &x, &y);
      graph[x + 1][y + 1] = 1;
    }

    ans = 0;

    for (int a = 1; a <= m; a++) {
      for (int b = 1; b <= n; b++) {
        if (graph[a][b]) {
          dfs(a, b);
          ans++;
        }
      }
    }

    printf("%d\n", ans);
  }

  return 0;
}
